std/algorithm/comparison.d

```diff
@@ -892,116 +892,367 @@ nothrow pure @safe unittest
+        {
+            static if (!is(typeof(allSameTypeIterative) == bool) && // not yet defined
+                       !is(V[0] == Vi)) // 10% faster than `!isSame(V[0], Vi)`
+            {
+                enum allSameTypeIterative = false;
+            }
+        }
+    }
+    static if (!is(typeof(allSameTypeIterative) == bool)) // if not yet defined
+    {
+        enum allSameTypeIterative = true;
+    }
+}
+
+/** Check if `pred` is a standard equality string or lambda predicate.
+ *
+ * See also: https://github.com/dlang/phobos/pull/3373#pullrequestreview-102956739
+ */
+private static template isEqualityPredicate(alias pred)
+{
+    static if (is(typeof(pred) == string))
+    {
+        // TODO remove when string lambdas become deprecated
+        enum isEqualityPredicate = (pred == "a == b");
+    }
+    else
+    {
+        enum isEqualityPredicate = __traits(isSame, pred, (a, b) => a == b);
+    }
+}
+
+/** Returns: `xs` forwarded through calls to `fun` and packed into a `std.typecons.Tuple`.
+ *
+ * See also: https://forum.dlang.org/post/zjxmreegqkxgdzvihvyk@forum.dlang.org
+ */
+private static auto forwardMap(alias fun, Ts...)(Ts xs) @trusted
+    if (is(typeof(unaryFun!(fun))))
+{
+    import std.meta : staticMap;
+    alias MappedTypeOf(T) = typeof(fun(T.init));
+    alias NewTypes = staticMap!(MappedTypeOf, Ts);
+
+    import std.typecons : Tuple;
+    Tuple!NewTypes ys = void;
+
+    alias fun_ = unaryFun!(fun);
+
+    import std.conv : emplace;
+    static foreach (immutable i, x; xs)
+    {
+        emplace(&ys[i], fun_(x));
+    }
+
+    return ys;
+}
+
 // equal
 /**
-Compares two ranges for equality, as defined by predicate `pred`
-(which is `==` by default).
+Compares two or more ranges `rs` for equality, as defined by predicate `pred`
+(which is `a == b` by default).
 */
 template equal(alias pred = "a == b")
 {
-    enum isEmptyRange(R) =
-        isInputRange!R && __traits(compiles, {static assert(R.empty, "");});
+    import std.meta : anySatisfy, allSatisfy;
+
+    /** Check if `T` is always empty (`empty` is an `enum` always being `true`)
+     */
+    static enum isAlwaysEmpty(T) = (__traits(compiles, { static assert(T.empty); }));
 
-    enum hasFixedLength(T) = hasLength!T || isNarrowString!T;
+    static enum areAllEquableRanges(Rs...) = is(typeof(binaryFun!pred(ElementType!(Rs[0]).init, // TODO check all `Rs`
+                                                                      ElementType!(Rs[1]).init)));
+
+    static alias ElementEncodingTypeUnqual(R) = Unqual!(ElementEncodingType!R);
+
+    static enum hasIndexingAndLength(T) = (is(typeof(T.init[0])) &&
+                                           is(typeof(T.length)));
+
+    /// TODO better name, anyone?
+    static auto maybeByCodeUnit(T)(T x)
+    {
+        static if (isAutodecodableString!T)
+        {
+            import std.utf : byCodeUnit;
+            return x.byCodeUnit;
+        }
+        else
+        {
+            return x;
+        }
```

Comment:
Wait isn't `byCodeUnit` a no-op for non auto-decodeable strings?
(and removing this entirely still compiles fine...)